### PR TITLE
[nextion] Use persistent http connection for TFT upload (ESP-IDF)

### PR DESCRIPTION
**Pull request by edwardtfn**
# What does this implement/fix?

TFT upload now transfers with a single stream, reducing the number of requests to the http server. In fact, this is similar to increasing the previously used chunk size to the file size, so all the data is requested to the http server in a single request, but then transferred to the UART on 4096 bytes chunks (as required by Nextion). If the display requests a different position, then the next request is the new http request (still in a persistent connection).
This will avoid a new http connection per chunk on Nextion when uploading TFT file, improving memory management and reliability of the whole process.
This is only for esp-idf. I will push a similar change to Arduino.
This is part of the breaking down of #6192

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** N/A

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

N/A

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/nextion/nextion_upload_idf.cpp [hunk 1]
```diff
@@ -20,152 +19,152 @@ static const char *const TAG = "nextion.upload.idf";
...
-        }
+    ESP_LOGD(TAG, "Successfully allocated upload buffer in PSRAM");
+  }
+#else
+  buffer = (uint8_t *) malloc(4096);
+#endif
+  if (!buffer) {
+    ESP_LOGE(TAG, "Failed to allocate upload buffer");
+    return -1;
+  }
```
Comment (jesserockz) [inline esphome/components/nextion/nextion_upload_idf.cpp]: Please see ExternalRAMAllocator which does all of this for you already

File: esphome/components/nextion/nextion_upload_idf.cpp [hunk 2]
```diff
@@ -249,68 +252,91 @@ bool Nextion::upload_tft() {
...
-  ESP_LOGD(TAG, "Restarting Nextion");
-  this->soft_reset();
-  vTaskDelay(pdMS_TO_TICKS(1500));  // NOLINT
+  this->ignore_is_setup_ = false;
+
   if (successful) {
     ESP_LOGD(TAG, "Restarting ESPHome");
-    esp_restart();  // NOLINT(readability-static-accessed-through-instance)
+    vTaskDelay(pdMS_TO_TICKS(1500));  // NOLINT
+    esp_restart();                    // NOLINT(readability-static-accessed-through-instance)
```
Comment (jesserockz) [inline esphome/components/nextion/nextion_upload_idf.cpp]:
```suggestion
    delay(1500);  // NOLINT
    arch_restart();
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please also move/duplicate `#define USE_NEXTION_TFT_UPLOAD` in `esphome/core/defines.h` so that the idf upload code is run by clang correctly. Currently it is not.

File: esphome/components/nextion/nextion_upload_idf.cpp (hunk 1 above)
Comment (edwardtfn) [reply esphome/components/nextion/nextion_upload_idf.cpp]: That makes things much easier. Thanks for pointing it.

**Comment by edwardtfn** [comment]
> Please also move/duplicate #define USE_NEXTION_TFT_UPLOAD in esphome/core/defines.h so that the idf upload code is run by clang correctly. Currently it is not.

I've changed that in PR #6661 as this wasn't duplicated by any of my previous PRs, so I wanna track this on the side.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**